### PR TITLE
CompletableFuture fixes + increased test coverage

### PR DESCRIPTION
**Pull request by Holmistr**
This is a round 1 of the CompletableFuture review.
* increased test coverage for CancellableDelegatingFuture
* minor cleanup
* fixed ClientDelegatingFuture caching (thanks @vbekiaris) and increased test coverage for it as well.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java [hunk 1]
```diff
@@ -382,6 +383,11 @@ public String toString() {
         return future.toString();
     }
 
+    // used for testing
+    V getDeserializedValue() {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java]: This method will be useful for testing of `DelegatingCompletableFuture` also, maybe it makes sense to pull it up?

File: hazelcast/src/test/java/com/hazelcast/client/impl/ClientDelegatingFutureTest.java [hunk 2]
```diff
@@ -176,6 +191,16 @@ public void thenCombine() {
         assertEquals(DESERIALIZED_VALUE, nextStage.join());
     }
 
+    @Test
+    public void thenCombineAsync() {
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/client/impl/ClientDelegatingFutureTest.java]: `CompletionStage` API methods of `ClientDelegatingFuture` are covered by `ClientDelegatingFutureTest_CompletionStageTest` which was introduced at a later point; probably it makes sense to remove all `CompletionStage` API tests from `ClientDelegatingFutureTest`?

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java (hunk 1 above)
Comment (Holmistr) [reply hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java]: The problem is that then I cannot use a package-private access modifier. Is it OK to make it public?

Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java]: I don't think it's a problem, this class is internal API.

**Comment by Holmistr** [comment]
@vbekiaris Comments addressed, please take another look :)

**Pull request merged**